docs/advanced/ecdsa.md

```diff
@@ -0,0 +1,46 @@
+
+### Protocol description
+
+The ECDSA authentication mechanism makes use of the SASL extension. This must be requested by the client prior to any authentication attempts (`CAP REQ :sasl`). Once the server sends an `ACK`, authentication can begin.
+
+The mechanism makes use of the NIST P-256 curve (`NID_X9_62_prime256v1` in OpenSSL) which (per the name) is defined over a 256 bit prime finite field. The parameters for this curve are available [from the NSA](https://www.nsa.gov/ia/_files/nist-routines.pdf) in section 4.3. The curve itself doesn't fit the [SafeCurves requirements](http://safecurves.cr.yp.to/).
+
+Atheme is the only IRC services software that appears to implement this authentication method. Atheme's implementation requires that:
+
+* Users make use of the `/msg NickServ SET PUBKEY <pubkey>` command to tell the services package their public key
+    * `<pubkey>` in this case is the **point compressed** X9.62 representation of the public key after it has been base64 encoded. Please see section 2.3.3 of the [SEC1](http://www.secg.org/sec1-v2.pdf) spec for more information. Public keys making use of the uncompressed or hybrid forms will **not work**. Java users may wish to implement their [own method](https://github.com/KittehOrg/KittehIRCClientLib/pull/46) for applying point compression since it is unavailable in the standard library.
+
+#### Step 1 - Begin authentication
+
+The client sends `AUTHENTICATE ECDSA-NIST256P-CHALLENGE` to request that the authentication process be started. See the [SASL documentation](http://ircv3.net/specs/extensions/sasl-3.1.html) for more information on the SASL process.
+
+### Step 2 - Wait for server acknowledgement
+
+The server will send back a `AUTHENTICATE +` message. The + here denotes an empty message. Once this message has been seen, the client can begin sending the user information.
+
+### Step 3 - Send encapsulated username information
+
+The username to authenticate with, along with the identity to impersonate are concatenated and joined together by \0 (NUL byte, U+0000) characters. For example, logging in as user Example and impersonating Test:
+
+`Example\0Test\0'
+
+In the vast majority of cases, the identity to impersonate will be identical to the identity used to authenticate:
+
+`Example\0Example\0`
+
+KICL does not support impersonation at this time and it appears to only be relevant to services administrators.
+
+Once the string has been built, it is base64 encoded and sent via the `AUTHENTICATE` command:
+
+`AUTHENTICATE RXhhbXBsZQBFeGFtcGxlAA==`
+
+### Step 4 - Receive challenge from server
+
+The server will now send a 32 byte challenge, encoded as base64. The client must **decode** the challenge (do not sign the base64 representation directly) and sign it, using the user's private key. The signature should be stored.
```

Comment:
Need to check if 32 bits or 32 bytes. 99.9% sure it's bytes since it matches the length of a sha256 hash (256 bits, 32 bytes).